src/common/HaskellDo/Toolbar/State.hs

```diff
@@ -47,21 +56,57 @@ update ClosePackageModal state = do
-    isProject <- atRemote $ localIO $ doesFileExist (projectPath newState ++ "package.yaml")
-    if isProject
-        then do
-            localIO $ setHtmlForId "#creationDisplay" ""
-            localIO $ setHtmlForId "#closeModalButton event .material-icons" "input"
-            return $ newState { createProject = False }
-        else do
-            localIO $ setHtmlForId "#creationDisplay" ("<p class=\"red-text\">No project found at " ++ projectPath newState ++ ", it will be created.</p>")
-            localIO $ setHtmlForId "#closeModalButton event .material-icons" "playlist_add"
-            return $ newState { createProject = True }
+    path <- pathOrLastOrHome newPath
+    localIO $ setValueForId "#pathInput event input" path
+
+    exists <- atRemote . localIO $ doesDirectoryExist path
+    let newState = state { directoryExists = exists, projectPath = path }
+
+    if exists
+    then do
+      (directories, files) <- directoriesAndFiles path
+
+      let newState' = newState { directoryList = (directories, files) }
+
+      isProject <- atRemote $ localIO $ doesFileExist (path </> "package.yaml")
+      updateProjectAvailability newState' path isProject
+    else do
+      let newState' = newState { directoryList = ([], []) }
+      return newState'
+  where
+    pathOrLastOrHome path = if null path
+                      then do
+                        exists <- atRemote . localIO $ doesFileExist lastProjectFile
+                        home <- atRemote . localIO $ getHomeDirectory
+                        -- if lastProjectFile exists and is not empty, use it
+                        -- otherwise use the home directory
+                        if exists
+                        then do
+                          content <- atRemote . localIO $ readFile lastProjectFile
+                          return $ if null content then home else content
+                        else
+                          return home
+                      else
+                        return path
+
+    directoriesAndFiles path = do
+      list <- atRemote . localIO $ listDirectory path
+      let visible = filter ((/= '.') . head) list
+      directories <- atRemote . localIO $ filterM (doesDirectoryExist . (path </>)) visible
+      files <- atRemote . localIO $ filterM (doesFileExist . (path </>)) visible
+
+      return (directories, files)
+
+    updateProjectAvailability currentState _ True = do
+      localIO $ setHtmlForId "#creationDisplay" ""
+      localIO $ setHtmlForId "#closeModalButton event .material-icons" "input"
+      return $ currentState { createProject = False }
+
+    updateProjectAvailability currentState path False = do
+      localIO $ setHtmlForId "#creationDisplay" ("<p class=\"red-text\">No project found at " ++ path ++ ", it will be created.</p>")
+      localIO $ setHtmlForId "#closeModalButton event .material-icons" "playlist_add"
+      return $ currentState { createProject = True }
```

Comment:
Yeah, it should use the new `updateDisplays` function to re-render the text content of the widget, just like the file tree. Another PR though, right?